apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.jbosslogmanager.correlation;
+
+import co.elastic.apm.agent.loginstr.correlation.AbstractLogCorrelationHelper;
+import org.jboss.logmanager.ExtLogRecord;
+
+public class JBossLogManagerCorrelationHelper extends AbstractLogCorrelationHelper.DefaultLogCorrelationHelper {
+
+    private final ThreadLocal<ExtLogRecord> cachedRecord = new ThreadLocal<>();
+
+    public boolean beforeLoggingEvent(ExtLogRecord record) {
+        cachedRecord.set(record);
+        return super.beforeLoggingEvent();
+    }
+
+    @Override
+    public void afterLoggingEvent(boolean addedToMdc) {
+        super.afterLoggingEvent(addedToMdc);
+        cachedRecord.remove();
+    }
+
+    @Override
+    protected void addToMdc(String key, String value) {
+        cachedRecord.get().putMdc(key, value);
+    }
+
+    @Override
+    protected void removeFromMdc(String key) {
+    }
```

Comment:
I wanted do that, but some adapters, e.g. the Log4J2 one, do call `ExtLogRecord#setMDC` when creating the `ExtLogRecord`. Since`ExtLogRecord` only "copies" the MDC once, that approach doesn’t work unfortunately.

Reply:
I see. Then in that case, I don't think we need to remove the IDs from the MDC. That will simplify things as you don't need to cache the record as a `ThreadLocal` - instead, you can add a hard-ref field to `JBossLogManagerCorrelationHelper` and `JBossLogManagerCorrelationHelper#beforeLoggingEvent()` can be implemented as follows:
```
    public boolean beforeLoggingEvent(ExtLogRecord record) {
        try {
            cachedRecord = record;
            return super.beforeLoggingEvent();
        } finally {
            cachedRecord = null;
        }
    }
```
That is safe enough.

Reply:
I don't think your approach will work when multiple threads log at the same time. As far as I can see `*CorrelationHelper` are shared by all threads so one thread might overwrite the cached record of another thread leading to incorrect correlation ids.

Reply:
Yes, you are absolutely right, it must be `ThreadLocal`, I don't know what I was thinking.
It works more elegantly with the other helpers because the event we instrumented allows direct update for the MDC, which implements the thread locality.
OK, then let's keep it this way.

Reply:
Yes, it is a bit unfortunate, that we cannot use the MDC here.

Reply:
We could also avoid using the abstract helper or change the API to allow passing along an Object (the log event in this case), but your `TheadLocal` approach seems to be the best.
Thanks!